Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,599 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+using MediatR;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.Notifications {
+  /// <summary>
+  /// DTO used to communicate the current verification diagnostics to the LSP client.
+  /// </summary>
+  [Method(DafnyRequestNames.VerificationDiagnostics, Direction.ServerToClient)]
+  public class VerificationDiagnosticsParams : IRequest, IRequest<Unit> {
+    public VerificationDiagnosticsParams(DocumentUri uri,
+      int version,
+      VerificationTree[] verificationTrees,
+      Container<Diagnostic> diagnostics,
+      int linesCount,
+      bool verificationStarted,
+      int numberOfResolutionErrors) {
+      Uri = uri;
+      Version = version;
+      VerificationTrees = verificationTrees;
+      if (linesCount != 0) { // Deserialization makes linesCount to be equal to zero.
+        PerLineDiagnostic =
+          RenderPerLineDiagnostics(this, verificationTrees, linesCount, numberOfResolutionErrors, verificationStarted, diagnostics);
+      }
+    }
+
+    /// <summary>
+    /// Gets the URI of the document whose verification completed.
+    /// </summary>
+    public DocumentUri Uri { get; init; }
+
+    /// <summary>
+    /// Gets the version of the document.
+    /// </summary>
+    public int? Version { get; init; }
+
+    /// <summary>
+    /// Returns a tree of diagnostics that can be used
+    /// First-level nodes are methods, or witness subset type checks
+    /// Second-level nodes are preconditions, postconditions, body verification status
+    /// Third level nodes are assertions inside functions 
+    /// </summary>
+    public VerificationTree[] VerificationTrees { get; init; }
+
+    /// <summary>
+    /// Returns per-line real-time diagnostic
+    /// </summary>
+    public LineVerificationStatus[] PerLineDiagnostic { get; init; }
+
+    static LineVerificationStatus[] RenderPerLineDiagnostics(
```

Comment:
I haven't been able to precisely understand what this method does, nor what the meaning of the output, the entries of the `LineVerificationStatus` enum, is. Would it be possible to include various example programs and their `RenderPerLineDiagnostics` output somewhere, either in test files or in diagnostics?

Reply:
Could you perhaps comment on the tests first?
https://github.com/dafny-lang/dafny/pull/1942/files#diff-245d3daf9d22b82772587d8de3b8a6c8097a5dbeeb6f5b58d77e0f0f3a3d2a8dR25

I created a format that should be straightforward: it display various `LineVerificationStatus[]` vertically, from left to right, to simulate what would happen in the IDE, with 3-char icons.
The same string is used both to define the program (after extracting all expected `LineVerificationStatus[]`), to perform some edits (using the ``//Next:` directive), and to test the successive `LineVerificationStatus[]` (after removing duplicates).

Let me know how if this helps.

Reply:
> Let me know how if this helps.

It's great! Really cool testing format. This is exactly what I was looking for. Sorry I must not have looked well before.